presto-main/src/main/resources/webapp/assets/utils.js

```diff
@@ -0,0 +1,387 @@
+            }
+            return "USER ERROR";
+        case "INTERNAL_ERROR":
+            return "INTERNAL ERROR";
+        case "INSUFFICIENT_RESOURCES":
+            return "INSUFFICIENT RESOURCES";
+        case "EXTERNAL":
+            return "EXTERNAL ERROR";
+    }
+    return errorType;
+}
+
+function getQueryStateColor(state, errorType, errorCode)
+{
+    switch (state) {
+        case "QUEUED":
+            return QUERY_STATE_COLOR_MAP.QUEUED;
+        case "PLANNING":
+        case "STARTING":
+        case "RUNNING":
+        case "FINISHING":
+            return QUERY_STATE_COLOR_MAP.RUNNING;
+        case "FAILED":
+            switch (errorType) {
+                case "USER_ERROR":
+                    if (errorCode.name === 'USER_CANCELED') {
+                        return QUERY_STATE_COLOR_MAP.USER_CANCELED;
+                    }
+                    return QUERY_STATE_COLOR_MAP.USER_ERROR;
+                case "EXTERNAL":
+                    return QUERY_STATE_COLOR_MAP.EXTERNAL_ERROR;
+                case "INSUFFICIENT_RESOURCES":
+                    return QUERY_STATE_COLOR_MAP.INSUFFICIENT_RESOURCES;
+                default:
+                    return QUERY_STATE_COLOR_MAP.UNKNOWN_ERROR;
+            }
+        case "FINISHED":
+            return QUERY_STATE_COLOR_MAP.FINISHED;
+        default:
+            return QUERY_STATE_COLOR_MAP.QUEUED;
+    }
+};
+
+// Sparkline-related functions
+// ===========================
+
+// display at most 5 minutes worth of data on the sparklines
+var MAX_HISTORY = 60 * 5;
+ // alpha param of exponentially weighted modving average. picked arbitrarily - lower values means more smoothness
+var EWMA_ALPHA = 0.2;
+
+function addToHistory (value, valuesArray) {
+    if (valuesArray.length == 0) {
+        return valuesArray.concat([value]);;
+    }
+    return valuesArray.concat([value]).slice(Math.max(valuesArray.length - MAX_HISTORY, 0));
+}
+
+function addExponentiallyWeightedToHistory (value, valuesArray) {
+    if (valuesArray.length == 0) {
+        return valuesArray.concat([value]);;
+    }
+
+    var ewma = (value * EWMA_ALPHA) + (valuesArray[valuesArray.length - 1] * (1 - EWMA_ALPHA));
+    if (value < 1) {
+        ewma = 0;
+    }
+
+    return valuesArray.concat([ewma]).slice(Math.max(valuesArray.length - MAX_HISTORY, 0));
+}
+
+// Utility functions
+// =================
+
+function getTaskIdSuffix(taskId) {
+    return taskId.slice(taskId.indexOf('.') + 1, taskId.length)
+}
+
+function getTaskIdInStage(taskId) {
+    return Number.parseInt(getTaskIdSuffix(getTaskIdSuffix(taskId)));
+}
+
+function removeQueryId(id)
+{
+    var pos = id.indexOf('.');
+    if (pos != -1) {
+        return id.substring(pos + 1);
+    }
+
+    return id
+}
+
+function taskIdToSortableString(id) {
+    var taskIdArr = removeQueryId(id).split(".");
+    return taskIdArr.map( function(numStr) { return new Array(5 - numStr.length).join("0") + numStr }).join("|");
```

Comment:
why 5? This formula seems a bit complicated.
Please add an example `id` so reader can understand this transformation